### ISSUE TITLE
possible memory leak on shift-drag / shift-click

### ISSUE THREAD

**Issue by tobwen** [Leaflet/Leaflet#6709]
- [X] I've looked at the [documentation](http://leafletjs.com/reference.html) to make sure the behavior is documented and expected
- [X] I'm sure this is a Leaflet code issue, not an issue with my own code nor with the framework I'm using (Cordova, Ionic, Angular, React…)
- [X] I've searched through the issues to make sure it's not yet reported

**Steps to reproduce**
1. shift + click into the map
2. check error messages in console

**Expected behavior**
Jump or zoom to the position.

**Current behavior**
This endlessly repeats and consumes memory:
```
Uncaught Error: Invalid LatLng object: (NaN, NaN)
    at new j (leaflet.js:5)
    at Object.unproject (leaflet.js:5)
    at Object.pointToLatLng (leaflet.js:5)
    at i.unproject (leaflet.js:5)
    at i.layerPointToLatLng (leaflet.js:5)
    at i._fireDOMEvent (leaflet.js:5)
    at i._handleDOMEvent (leaflet.js:5)
    at HTMLDivElement.s (leaflet.js:5)
```

**Environment**
- Leaflet version: 1.5.1
- Browser (with version): Chrome 75.0.3770.100
- OS/Platform (with version): Microsoft Windows 7 Professional, Version 6.1.7601

**Additional context**
Sure, the extent is pretty, but I believe it's an unhandled bug.

**Minimal example reproducing the issue**
https://codepen.io/1100101/pen/wLyxQY

- [X] this example is as simple as possible
- [X] this example does not rely on any third party code

**Comment by IvanSanchez** [comment]
Cannot reproduce. Nothing on my console when shift-clicking or shift-dragging a 1px box (using Chrome 75 on Debian and the linked codepen).

**Comment by tobwen** [comment]
Can be reproduced on 3 different Windows machines: 2x Windows 10, 1x Windows 7.
Does NOT appear on Firefox.

I'll add a screencast for you in some minutes.

**Comment by tobwen** [comment]
![leaflet_shift_klick](https://user-images.githubusercontent.com/1864057/60467977-6bbb4180-9c58-11e9-8554-10f110c8051c.gif)

**Comment by tobwen** [comment]
@IvanSanchez Can you have a look please? It's definitely not imagination. I've setup a fresh Chrome profile without any extensions (addons) and it's still reproducible.

**Comment by IvanSanchez** [comment]
So this is windows-specific, then? Weird. I'll reopen, though right now I don't have a windows box to test this.

@tobwen Please use `leaflet-src.js` instead of `leaflet.js`, you'll have better stack traces that way.

**Comment by tobwen** [comment]
@IvanSanchez I've updated the pen to use `leaflet-src.js` and here's the error log:

```
LatLng.js:32 Uncaught Error: Invalid LatLng object: (NaN, NaN)
    at new LatLng (LatLng.js:32)
    at Object.unproject (Projection.SphericalMercator.js:35)
    at Object.pointToLatLng (CRS.js:41)
    at NewClass.unproject (Map.js:999)
    at NewClass.layerPointToLatLng (Map.js:1007)
    at NewClass._fireDOMEvent (Map.js:1435)
    at NewClass._handleDOMEvent (Map.js:1396)
    at HTMLDivElement.handler (DomEvent.js:79)
```

**Comment by IvanSanchez** [comment]
BTW are you using a hi-dpi screen for this?

**Comment by tobwen** [comment]
> BTW are you using a hi-dpi screen for this?

Nope, 1680 x 1050 on my office desktop, but also HD ones (just tested). I don't have hi-dpi/retina to test :(

**Comment by qligier** [comment]
This bug happens when the zoom level reaches Infinity (the default limit). `_getNewPixelOrigin()` returns `Point {x: Infinity, y: Infinity}` (probably because of the point projection at Infinity zoom level). Then, every call to `layerPointToLatLng()` returns the same Infinity point and nothing works anymore.

You can set the `maxZoom` property of the map to prevent it.

**Comment by ryanmoore-SM** [comment]
> This bug happens when the zoom level reaches Infinity (the default limit). `_getNewPixelOrigin()` returns `Point {x: Infinity, y: Infinity}` (probably because of the point projection at Infinity zoom level). Then, every call to `layerPointToLatLng()` returns the same Infinity point and nothing works anymore.
> 
> You can set the `maxZoom` property of the map to prevent it.

Just chiming in to say I wrestled with this problem for a couple days before I found this comment. Apparently a value of 9999 for maxZoom is not valid. Setting it to 999 fixed the `Invalid LatLng object: (NaN, NaN)` errors

**Comment by johnd0e** [comment]
Unable to reproduce (on Windows machine)

**Comment by qligier** [comment]
Shift-clicking doesn't work anymore on my side either, but it's still reproducible by holding shift and zooming by selecting small rectangles on the map, multiple times (between 1 and 10 times are usually enough, depending on the size of the rectangles). I'm using Chrome 95 on a Windows 10 machine.

**Comment by Falke-Design** [comment]
Yep reproduced 👍  the important part: "multiple times"

**Comment by johnd0e** [comment]
The error caused by `zoom` = `Infinity`, but now we need to track down where it gets such value..

**Comment by johnd0e** [comment]
Ok, tracked that down.
The issue arises (in my case) on zoom 54, when different points unprojects to the same latlng value,
Here we get Infinity:
https://github.com/Leaflet/Leaflet/blob/a37f0297afd1e547e94d5ae4e217f7cef1d696c8/src/map/handler/Map.BoxZoom.js#L131-L133

**Comment by johnd0e** [comment]
Zoom becomes `Infinity` because `boundsSize` are [0,0] here.
https://github.com/Leaflet/Leaflet/blob/a37f0297afd1e547e94d5ae4e217f7cef1d696c8/src/map/Map.js#L870

While the root of the issue is clear know, i have no idea in what place should that be fixed..
@IvanSanchez @mourner

**Comment by IvanSanchez** [comment]
@johnd0e This happens only when the map doesn't have a `maxZoom` option set, right? (and no tilelayers/gridlayers with `maxZoom` either).

We could add an extra sanity check at `Map.BoxZoom.js` - instead of creating the `LatLngBounds`, store the start and end points in two variables, and run `L.LatLng.equals()` (https://leafletjs.com/reference.html#latlng-equals) on them.

**Comment by johnd0e** [comment]
@IvanSanchez May be better fix that in [`_getBoundsCenterZoom`](https://github.com/Leaflet/Leaflet/blob/a37f0297afd1e547e94d5ae4e217f7cef1d696c8/src/map/Map.js#L260-L269)?

**Comment by IvanSanchez** [comment]
> @IvanSanchez May be better fix that in [`_getBoundsCenterZoom`](https://github.com/Leaflet/Leaflet/blob/a37f0297afd1e547e94d5ae4e217f7cef1d696c8/src/map/Map.js#L260-L269)?

I disagree - the goal of `_getBoundsCenterZoom` is to take a `LatLngBounds`, and return its center and a zoom level, and it's OK if it returns `Infinity` as per 873d4ca / #5157 / #5153. 

Right now zero-area `LatLngBounds` are a problem only when box-zooming. If there is any other way to reproduce a similar bug (other than box-zooming), then I would agree to tweak `_getBoundsCenterZoom` instead.

**Comment by johnd0e** [comment]
> If there is any other way to reproduce a similar bug [...]

As per test in #7771 - direct call of `fitBounds` with zero-sized bounds. Whatever reason.

**Comment by johnd0e** [comment]
Alternatively, we'd could protect `getBoundsZoom` from returning `Infinity`, unless you know some case when it can be useful.

**Comment by johnd0e** [comment]
Alternatively, we'd could:
-  Either [protect](https://github.com/johnd0e/Leaflet/commit/059b783e71df75e774465abc913488977ae82038) `getBoundsZoom` from returning `Infinity`, unless you know [some case](https://leafletjs.com/reference.html#map-getboundszoom) when it can be useful.
- Or [fix](https://github.com/johnd0e/Leaflet/commit/8873a6432cefe41ff4deff95c9cbf764c5bc6f3e) `fitBounds` itself.
